### DISCUSSION TITLE
Vue Directives are not Typed making usage hard in TS environments

### DISCUSSION THREAD

**Discussion by HarisSpahija** [quasarframework/quasar#15798] [Ideas / Proposals] [answered]
Currently there are no interfaces or types present in the vue directives. Suggestion is to convert https://github.com/quasarframework/quasar/tree/dev/ui/src/directives and the existing .js files to either typescript or generate a type definition file for accessing the interfaces for the payloads.

For example TouchPan.js should have an interface to be imported when using the event in a method using the directive

```js
// TouchPan.js
 return {
    synthetic,
    payload: {
      evt,
      touch: ctx.event.mouse !== true,
      mouse: ctx.event.mouse === true,
      position: pos,
      direction: dir,
      isFirst: ctx.event.isFirst,
      isFinal: isFinal === true,
      duration: Date.now() - ctx.event.time,
      distance: {
        x: absX,
        y: absY
      },
      offset: {
        x: distX,
        y: distY
      },
      delta: {
        x: pos.left - ctx.event.lastX,
        y: pos.top - ctx.event.lastY
      }
    }
  }
  ```
Current Usage
```
<template>
  <div v-touch-pan.prevent.mouse="touchAction">
    Hello World
  </div>
</template>  
<script setup lang="ts">
const touchAction = (event: any) => {}
</script>
```
Expected Usage
```
<template>
  <div v-touch-pan.prevent.mouse="touchAction">
    Hello World
  </div>
</template>  
<script setup lang="ts">
import { TouchPanEvent } from "path/to/typedefinition"
const touchAction = (event: TouchPanEvent) => {}
</script>
```

**Comment by HarisSpahija** [comment]
I can knock this out really quick, just want to make sure its inline with the direction of the directives to translate to Typescript or create interfaces

**Comment by yusufkandemir** [comment, marked as answer]
This was already planned, with a bigger scope. Here is the implementation PR: https://github.com/quasarframework/quasar/pull/16068
